### PR TITLE
Refactor Callback Function in generate.py

### PR DESCRIPTION
**Pull request by neverbiasu**
This PR refactors the _callback function in generate.py. Splitting `_callback` into `_callback`, `_check_done_generating`.

### CODE REVIEW

**Review by byjlw — CHANGES_REQUESTED** [review]
Basic command from the readme fails. Please test your change against the instructions in the readme to ensure things work correctly. 

```
(.venv) byjlw@byjlw-mbp torchchat % python3 torchchat.py generate llama3.1 --prompt "write me a story about a boy and his bear"

Warning: compilation is not available with device MPS, ignoring option to engage compilation
NumExpr defaulting to 10 threads.
PyTorch version 2.5.0.dev20240710 available.
Using device=mps 
Loading model...
Time to load model: 25.84 seconds
-----------------------------------------------------------
Traceback (most recent call last):
  File "/Users/byjlw/Documents/scrap/torchchat/torchchat.py", line 88, in <module>
    generate_main(args)
  File "/Users/byjlw/Documents/scrap/torchchat/generate.py", line 850, in main
    for _ in gen.chat(generator_args):
  File "/Users/byjlw/Documents/scrap/torchchat/generate.py", line 765, in chat
    for token_tensor, metrics in generator_func:
  File "/Users/byjlw/Documents/scrap/torchchat/.venv/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 57, in generator_context
    response = gen.send(request)
               ^^^^^^^^^^^^^^^^^
  File "/Users/byjlw/Documents/scrap/torchchat/generate.py", line 540, in generate
    for generated_token, _ in self.decode_n_tokens(
  File "/Users/byjlw/Documents/scrap/torchchat/generate.py", line 338, in decode_n_tokens
    callback(new_tokens[-1], done_generating=_i == num_new_tokens - 2)
  File "/Users/byjlw/Documents/scrap/torchchat/generate.py", line 733, in callback
    return self._callback(
           ^^^^^^^^^^^^^^^
  File "/Users/byjlw/Documents/scrap/torchchat/generate.py", line 581, in _callback
    self._output(buffer)
    ^^^^^^^^^^^^
AttributeError: 'Generator' object has no attribute '_output'
```

File: generate.py [hunk 1]
```diff
@@ -572,23 +572,22 @@ def encode_tokens(self, string, bos=True, device="cpu"):
...
-            self.is_llama3_model
-            and x.item() == self.tokenizer.special_tokens["<|eot_id|>"]
-        ):
-            done_generating = True
-            buffer = buffer[:-1]  # drop the eot_id from the output buffer
+        )
+        done_generating = self._check_done_generating(x, buffer)
         if len(buffer) == 4 or done_generating:
-            print("".join(buffer), end="", flush=True)
+            self._output(buffer)
```
Comment (neverbiasu) [inline generate.py:581]: it should be removed, `self._output `used to be print.

Comment (neverbiasu) [inline generate.py:581]:
```suggestion
```

**Comment by Jack-Khuu** [comment]
Closing stale PR

**Pull request closed without merging**